security/advancedtls/crl.go

```diff
@@ -83,13 +88,46 @@ func (s RevocationStatus) String() string {
 	return [...]string{"RevocationUndetermined", "RevocationUnrevoked", "RevocationRevoked"}[s]
 }
 
-// certificateListExt contains a pkix.CertificateList and parsed
-// extensions that aren't provided by the golang CRL parser.
-type certificateListExt struct {
-	CertList *x509.RevocationList
+// CRL contains a pkix.CertificateList and parsed extensions that aren't
+// provided by the golang CRL parser.
+// All CRLs should be loaded using NewCRL() for bytes directly or ReadCRLFile()
+// to read directly from a filepath
+type CRL struct {
+	certList *x509.RevocationList
 	// RFC5280, 5.2.1, all conforming CRLs must have a AKID with the ID method.
-	AuthorityKeyID []byte
-	RawIssuer      []byte
+	authorityKeyID []byte
+	rawIssuer      []byte
+}
+
+// NewCRL constructs new CRL from the provided byte array.
+func NewCRL(b []byte) (*CRL, error) {
+	crl, err := parseRevocationList(b)
+	if err != nil {
+		return nil, fmt.Errorf("parseCrl() failed err = %v", err)
+	}
+	crlExt, err := parseCRLExtensions(crl)
+	if err != nil {
+		return nil, fmt.Errorf("parseCRLExtensions() failed err = %v", err)
+	}
+	crlExt.rawIssuer, err = extractCRLIssuer(b)
+	if err != nil {
+		return nil, fmt.Errorf("extractCRLIssuer() failed err= %v", err)
+	}
+	return crlExt, nil
+}
+
+// ReadCRLFile reads a file from the provided path, and returns constructed CRL
+// struct from it.
+func ReadCRLFile(path string) (*CRL, error) {
+	b, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("readFile(%v) failed err = %v", path, err)
```

Comment:
`os.ReadFile`?

Also I don't think this is the right format for errors in Go.  This is more the style for test failures, as it appears to be suited for programmers and not users.  Errors would usually look more like:

```go
return nil, fmt.Errorf("failed to read %q: %v", path, err)
```

(or `%w` instead of `%v` if you want users to be able to unwrap the read error, and then this becomes part of your API.)

You can see examples from the stdlib:

https://github.com/search?q=repo%3Agolang%2Fgo%20fmt.Errorf&type=code

Reply:
That original line was about what you were printing.  `readFile()` wasn't the thing that failed (and it doesn't exist); it was `os.ReadFile` that failed.  So it would be better to use the proper name of the thing that failed and not a function that doesn't exist.  But it's even better to not use debugging-style messages as errors, so then I suggested that.

Reply:
Oh I see, sorry - GitHub always highlights few lines so I thought there might be a better than `os.ReadFile` function to use here.